test/4_permissions_test.js

```diff
@@ -102,6 +118,38 @@ contract('LinniaPermissions', accounts => {
         })
       );
     });
+    it('should not allow non-delegate to grant access', async () => {
+      const fakeIpfsHash = eutil.bufferToHex(crypto.randomBytes(32));
+      await assertRevert(
+        instance.grantAccessbyDelegate(testDataHash1, provider2, user1, fakeIpfsHash, {
+          from: delegate
+        })
+      );
+      await assertRevert(
+        instance.grantAccessbyDelegate(testDataHash2, provider1, provider2, fakeIpfsHash, {
+          from: delegate
+        })
+      );
+    });
+    it('should allow delegate to grant access', async () => {
+      const fakeIpfsHash = eutil.bufferToHex(crypto.randomBytes(32));
+      await instance.addDelegate(delegate, {from: user1});
+      const tx = await instance.grantAccessbyDelegate(
+        testDataHash1,
+        provider2,
```

Comment:
It is a bit confusing why you are granting access to a `provider`.  Consider renaming `viewer`.